### PR TITLE
adding np.resize to the buffer to allow partial filling

### PR DESCRIPTION
**Pull request by kalwalt**
I added the resize feature to the texture viz node as the lite version does. As @Kosvor2  did in the lite version i changed the input buffer, not real a big effort to do this but hadn't the time before to code and to test it.
@zeffii Please test and review it,  for me it's ok but check if it works in other Os.  Merge it when you want.

### CODE REVIEW

**Comment by kalwalt** [comment]
This works for displaying in the node tree but will not works if we want to pass the data to the internal blender image editor. Do you want also this feature?

**Comment by kalwalt** [comment]
with the latest commit allow partial filling also for **BW** mode, **RGB** and **RGBA** is not allowed, need to figure out how to do this.

**Comment by zeffii** [comment]
i can't push this right now

**Comment by kalwalt** [comment]
no worries @zeffii  i will do this in my spare time, i will see for the **RGB RGBA** part...

**Comment by zeffii** [comment]
cool! take your time

File: nodes/viz/texture_viewer.py [hunk 1]
```diff
@@ -369,7 +369,7 @@ def process(self):
             pixels = np.array(self.inputs['Float'].sv_get(deepcopy=False)).flatten()
             width, height = self.texture_width_height
 
-            transfer_to_image(np.resize(pixels, self.total_size), self.texture_name, width, height, mode)
+            transfer_to_image(np.resize(pixels, self.calculate_total_size()), self.texture_name, width, height, mode)
```
Comment (kalwalt) [inline nodes/viz/texture_viewer.py]: this allow **rgb** and **rgba** mode but i get some 'red' alert ( i mean the texture viz node became red)

**Comment by kalwalt** [comment]
stopping for now because too late in the night, i will see in the next days this issue...

**Comment by kalwalt** [comment]
this from the console:
```
Traceback (most recent call last):
  File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/data_structure.py", line 447, in updateNode
    self.process_node(context)
  File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/node_tree.py", line 485, in process_node
    process_from_node(self)
  File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/core/update_system.py", line 423, in process_from_node
    do_update(update_list, nodes)
  File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/core/update_system.py", line 349, in do_update
    do_update_general(node_list, nodes)
  File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/core/update_system.py", line 313, in do_update_general
    done_nodes = set(procesed_nodes)
RecursionError: maximum recursion depth exceeded while calling a Python object
File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/data_structure.py", line 442, in updateNode

```

**Comment by zeffii** [comment]
are you sure you pushed all commits? I don't see any reason for recursion in the current code..

**Comment by zeffii** [comment]
it might be that the update into the Image is causing a scene update, which in turn causes a sverchok update..

**Comment by kalwalt** [comment]
>  are you sure you pushed all commits? I don't see any reason for recursion in the current code..

yes, i'm sure.

> it might be that the update into the Image is causing a scene update, which in turn causes a sverchok 
update.

i will look into it when i have a bit of time.

**Comment by kalwalt** [comment]
The recursion error appear when in you set the the dimension of the image to user(custom) and if width or heigth go to zero and re-assign a value greater than zero. The data are passed to the image viewer but the node stays in the `red` state.

**Comment by kalwalt** [comment]
How can avoid this?

**Comment by kalwalt** [comment]
min for width and heigth are set to 0 i will set to 1, i will see if this make a difference...

**Comment by kalwalt** [comment]
The simplest and obvious solutions are always the best ones!

**Comment by kalwalt** [comment]
...and also min = 0 does not make any sense to me...!

File: nodes/viz/texture_viewer.py [hunk 2]
```diff
@@ -369,7 +369,7 @@ def process(self):
             pixels = np.array(self.inputs['Float'].sv_get(deepcopy=False)).flatten()
             width, height = self.texture_width_height
 
-            transfer_to_image(pixels, self.texture_name, width, height, mode)
+            transfer_to_image(np.resize(pixels, self.calculate_total_size()), self.texture_name, width, height, mode)
```
Comment (kalwalt) [inline nodes/viz/texture_viewer.py]: @zeffii maybe better to put in two lines this?
```python
            transfer_to_image(np.resize(pixels, self.calculate_total_size()),
                              self.texture_name, width, height, mode)
```

Comment (zeffii) [inline nodes/viz/texture_viewer.py]: or
```
        resized_np_array = np.resize(pixels, self.calculate_total_size())
        transfer_to_image(resized_np_array, self.texture_name, width, height, mode)
```

**Comment by kalwalt** [comment]
i get again that error while increasing the size of the tex in the custom mode:
```
Traceback (most recent call last):
  File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/data_structure.py", line 447, in updateNode
    self.process_node(context)
  File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/node_tree.py", line 485, in process_node
    process_from_node(self)
  File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/core/update_system.py", line 423, in process_from_node
    do_update(update_list, nodes)
  File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/core/update_system.py", line 349, in do_update
    do_update_general(node_list, nodes)
  File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/core/update_system.py", line 313, in do_update_general
    done_nodes = set(procesed_nodes)
RecursionError: maximum recursion depth exceeded while calling a Python object
File "/home/walter/build_linux/bin/2.78/scripts/addons/sverchok/data_structure.py", line 442, in updateNode
```
but at least the node now not became red.

**Comment by kalwalt** [comment]
maybe we should do something like this:
```python
# inside Texture viever class

def update_to_image():
     mode = self.color_mode
     self.activate = False
     pixels = np.array(self.inputs['Float'].sv_get(deepcopy=False)).flatten()
     width, height = self.texture_width_height
     resized_np_array = np.resize(pixels, self.calculate_total_size())
     transfer_to_image(resized_np_array, self.texture_name, width, height, mode)
     updateNode(self, context)

if self.to_image_viewer:
    self.update_to_image()
```

**Comment by kalwalt** [comment]
`updateNode` is intented only for the enum properties?

**Comment by kalwalt** [comment]
the `update_to_image` solution doesn't works of course, but i found that this line of code is causing the 'recursion' issue : 
            `self.activate = False`
 https://github.com/kalwalt/sverchok/blob/b69a429774bdbeca2f651ae372897342d4553025/nodes/viz/texture_viewer.py#L368

we can simply delete it or find another solution but i  have any idea....

**Comment by zeffii** [comment]
then wrap updateNode for that scenario
```
    def wrapped_updateNode_(self, context):
        self.activate = False
        self.process()

    to_image_viewer = BoolProperty(
        name='Pass', description='Transfer pixels to image viewer',
        default=False, update=wrapped_updateNode_)
```

**Comment by zeffii** [comment]
meaning it doesn't need to be set from the process function .. (which is whats causing the retrig as you point out)

**Comment by zeffii** [comment]
or just
```
    def wrapped_updateNode_(self, context):
        self.activate = False
        # self.process()
```

**Comment by zeffii** [comment]
changing self.activate will trigger `self.process` anyway, so that is safe to drop from the wrapper..

**Comment by zeffii** [comment]
and ofcourse remove that line to `self.activate = False` from the process function itself.

**Comment by kalwalt** [comment]
i think this will solve the issue... i'm trying right now...

**Comment by kalwalt** [comment]
yes! but for safety i will check sunday, since i will have any time tomorrow.... thanks @zeffii !

**Comment by kalwalt** [comment]
but shoul i add also updateNode(self, context) in the wrapped update function?

**Comment by zeffii** [comment]
nope.. because  if you modify `self.activate` from inside the `wrapped_updateNode_` , that action will still call updateNode.. because that's the update function of `self.activate`

**Comment by zeffii** [comment]
the point of removing the explicit call to `self.activate=False` from inside the process function is to prevent the loop 

```
    def process
    
        self.activate = X   # this calls self.process again
```

**Comment by zeffii** [comment]
good catch btw.

**Comment by kalwalt** [comment]
understood. 

> good catch btw.

I went for exclusion, with a bit of lucky...!

**Comment by kalwalt** [comment]
i should do also another refinement in the code and after i think this will be fine.

**Comment by kalwalt** [comment]
I'm fine with it. What do you think @zeffii?

**Pull request merged**